### ISSUE TITLE
[Feature Request] Improve support for package version

### ISSUE THREAD

**Issue by leoafarias** [pattobrien/cli-gen#6]
Hi @pattobrien, first excellent job in the package; this looks good! 

I might be adding some issues or requests here based on common CLI problems that might fall under the scope of this package.

## Problem
Make the auto generation of the CLI version and command more effortless and seamless.

## Description
CLIs need to easily check the version. This is used for update checking, compatibility, user support, and so on. Right now, most have to use [build_version](https://pub.dev/packages/build_version) or a custom script (my case) in order to write the version value from pubspec.

## Request
- cli_gen should automatically generate a version.g.dart  file based on the pubspec.
- `@cliRunner` should allow for top level options similar to the `@cliCommand`, so we can have `$ cmd --version` which is "somewhat" the standard.

The top level arguments its important on the runner as that is the way also we can configfure global arguments for all commands, or log levels like `--verbose` and so on.

**Comment by pattobrien** [comment]
Hey @leoafarias ! 

I'm excited that you were able to give the package a try, and really appreciate the well thought out feedback! Please keep the issues / feature requests coming, and I'll do my best to get those out the door as quickly as possible! :)

I'll break this issue down into two main features: Built-in support for versioning, and handling globally-scoped args.

## Built-in Support for Versions

I think it absolutely makes sense to build this functionality directly into `cli_gen`. I'm wondering if there's any reason you rolled your own tool - just in case there's some use case `build_version` is missing that I'm not seeing.

Otherwise, to make the UX a little better, it may make the most sense for `cli_gen` to generate a public `version` constant in the same (part) file as the `Runner` so that we don't need to generate an extra `version.g.dart` file - what do you think?

## Top Level Options

I'm breaking this one down in 2 parts: a) option declarations and b) argument handling.

From my look over the `fvm` repo, from what I can tell, global args are only consumed by the Runner, would you say that's more or less correct?

### Proposed Solution(s)

There are a couple good ways we could have users declare and consume global args:

- 1: create a Runner method for handling globals, via a dedicated annotation (e.g. `@cliRun()`, `@cliEntrypoint`, etc) 
- 2: create a Runner method for handling globals, reserving a namespace like `runApp()` without an annotation
- 3: allow any `Runner`, `Subcommand` or `Command` class/method to be wrapped by a middleware that handles arg parsing before methods are called (more on this below)

In options 1 & 2, `cli_gen` would generate options/flags just as it already does with `@cliCommand`s, so the following would generate `--verbose` and `--version` flags:

```dart
@cliRunner
class Runner {
  @override
  void runApp({
    bool isVerbose = false,
    bool version = false,
  }) { 
   /* ... */ 
  }
}
```

Option 3 is at least a few months from being shippable, but I've been experimenting with such a DI system on another backend-related project I'm working on in parallel to `cli_gen`. Essentially it would look something like the following:

```dart
@use({
  loggerMiddleware,
  versionMiddleware,
})
@cliRunner
class Runner {
  ...
}
```

In short, any middleware on a `cli_gen` annotated class would run before any of the methods on the enclosing class. For the logger middleware, it would set the `logger` to `verbose` and then continue on, whereas the version middleware would handle the entire request itself.

For the examples you gave, Option 3 obviously seems like overkill, but for features like authentication or general configuration setup, you could see how this could be a really nice quality of life improvement.

### Suggestion

I think we should move forward with option 2, which would be easiest to implement in the near term and would handle the two use cases you mentioned.

In the future, if/when a DI / middleware system is possible, then I think that would be the most reasonable way to go long-term, as it allows for unlocking more use cases that Option 2 can on its own.

**Comment by leoafarias** [comment]
For the version generation, I didn't create a "tool." I have grinder scripts to do specific tasks, so it didn't make sense to have build_runner just for the version generation.

Option 2 makes sense. FYI, this is not a blocker; I got the version by overriding the `runCommand` and adding the "version" through the argParser like this.

```dart
@CliRunner(
  name: packageName,
  description: description,
)
class SuperDeckRunner extends _$SuperDeckRunner<int> {
  SuperDeckRunner() {
    argParser.addFlag('version', abbr: 'v', help: 'Print the version');
  }

  /// Builds the slides configured within the project
  @cliCommand
  Future<int> build({
    @Flag(
      abbr: 'w',
      help: 'Watch for changes to the slides.md file',
      negatable: false,
    )
    bool watch = false,
  }) {
    return buildCommandRunner(watch: watch);
  }

  @override
  Future<int?> runCommand(ArgResults topLevelResults) async {
    if (topLevelResults['version'] == true) {
      logger.info(packageVersion);
      return ExitCode.success.code;
    }
    try {
      return await super.runCommand(topLevelResults);
    } on UsageException catch (e) {
      stdout.writeln('${e.message}\n');
      stdout.writeln(e.usage);
    }
    return null;
  }
}
```

Also, a CLI tool like FVM might require more work than its worth to fit within the cli_gen. The global variable problem is done through dependency injection of a scope context to make it easier for testing and so on. I think cli_gen can focus on the DX of a CLI but only add a little magic to a lot that is going on. 

Middleware might make sense in certain use cases but will probably not resolve the more advanced use cases, and probably it shouldn't.  For catching errors and exceptions might be a great use case of it.

One last thing: If you are following the macros feature for Dart, it is still being prepared, but it seems it's progressing well. It might be worth following it closely as it might help many of the things you are currently working on.

**Comment by pattobrien** [comment]
> I got the version by overriding the runCommand and adding the "version" through the argParser like this.

I spent a lot of time designing `cli_gen` to allow access to the underlying APIs, so that's really assuring to hear!

> will probably not resolve the more advanced use cases, and probably it shouldn't. 

That's a good point, I hadn't thought of it that way. There's still a good amount of experimentation that I planned to see if such a system would work or not, but your experience holds more weight tbh.

> If you are following the macros feature for Dart...

I absolutely have been! The first iteration of this package was actually made with macros a couple weeks ago (as much as could be done at the time). Unless things really change, there should be almost no differences between the current implementation of cli_gen vs. with macros; and this is the reason why Parameter doc comments currently don't work (because its unclear if the same workaround will be doable using macros).

**Comment by pattobrien** [comment]
@leoafarias hey Leo! Just to let you know, I just published v0.1.0-dev.6 that includes automatically inferring `version` values from `pubspec.yaml`, and automatic generation of a `--version` command.

I'll close this issue as done! Thanks for filing :)
